src/keys/slip10.rs

```diff
@@ -44,100 +99,404 @@ impl Seed {
+            if !key.is_public_key_valid() {
+                Err(crate::Error::InvalidArgumentError {
+                    alg: "SLIP10",
+                    expected: "valid extended public key",
+                })
+            } else {
+                Ok(Self(key))
+            }
+        }
+    }
+
+    pub fn curve(&self) -> Curve {
+        self.0.curve
+    }
+
+    pub fn extended_bytes(&self) -> &[u8; 65] {
+        self.0.extended_public_bytes()
+    }
+
+    pub fn public_bytes(&self) -> &[u8; 33] {
+        self.0.public_bytes()
+    }
 
-        Ok(Self(i))
+    pub fn public_key(&self) -> PublicKey {
+        match self.curve() {
+            #[cfg(feature = "ed25519")]
+            Curve::Ed25519 => {
+                unreachable!("SLIP10 does not support non-hardened key derivation from Ed25519 public keys")
+            }
+            #[cfg(feature = "secp256k1")]
+            Curve::Secp256k1 => secp256k1_ecdsa::PublicKey::try_from_bytes(self.public_bytes())
+                .map(PublicKey::Secp256k1Ecdsa)
+                .expect("valid extended public key"),
+        }
+    }
+
+    pub fn chain_code(&self) -> &[u8; 32] {
+        self.0.chain_code()
+    }
+
+    pub fn child_key(&self, segment: &Segment) -> crate::Result<Self> {
+        #[cfg(feature = "ed25519")]
+        debug_assert_ne!(Curve::Ed25519, self.curve());
+
+        if !segment.is_hardened() {
+            Ok(Self(self.0.child_key(segment)))
+        } else {
+            Err(crate::Error::InvalidArgumentError {
+                alg: "SLIP10",
+                expected: "non hardened key index for master public key",
+            })
+        }
     }
 
-    pub fn derive(&self, chain: &Chain) -> crate::Result<Key> {
-        let mut k = *self;
-        for c in &chain.0 {
-            k = k.child_key(c)?;
+    pub fn derive(&self, chain: &Chain) -> crate::Result<Self> {
+        #[cfg(feature = "ed25519")]
+        debug_assert_ne!(Curve::Ed25519, self.curve());
+
+        if chain.all_non_hardened() {
+            let mut key = self.0.clone();
+            for segment in &chain.0 {
+                key = key.child_key(segment);
+            }
+            Ok(Self(key))
+        } else {
+            Err(crate::Error::InvalidArgumentError {
+                alg: "SLIP10",
+                expected: "non hardened key index for master public key",
+            })
         }
-        Ok(k)
     }
 }
 
-impl TryFrom<&[u8]> for Key {
+impl TryFrom<&ExtendedSecretKey> for ExtendedPublicKey {
     type Error = crate::Error;
+    fn try_from(esk: &ExtendedSecretKey) -> crate::Result<Self> {
+        esk.try_into_extended_public_key()
+    }
+}
 
-    fn try_from(bs: &[u8]) -> Result<Self, Self::Error> {
-        if bs.len() != 64 {
-            return Err(crate::Error::BufferSize {
-                name: "key",
-                has: bs.len(),
-                needs: 64,
-            });
+#[derive(Clone)]
+struct KeyImpl {
+    curve: Curve,
+    ext: [u8; 65],
```

Comment:
This is secret/public key + chain code ? Maybe add a small comment about the content of the bytes

Reply:
Done.